### ISSUE TITLE
data corruption after synchronization to qgis

### ISSUE THREAD

**Issue by elmonchas** [opengisch/QField#5081]
### Describe the issue
I've noticed an issue with our data collection project. The issue arises when splitting and merging polygon features. The splitting and merging actions result in one polygon in qfield. After downloading the data through qfield cloud, the polygon overlaps whith the previously merged snippets. 

 
![image](https://github.com/opengisch/QField/assets/158583631/3989ac3b-80ee-4abb-b419-f268cc818093)

Please see screenshot above: The two red features were cut from the neigbour and merged with polygon "O30".  Polygon O30 overlaps in qgis with the two red snippets.  In qfield app the two red snippets dont exist as they were merged with O30.

I am suspicous the data corruption occurs due to changes in FID values. 
the FID (in qgis) of the two small errenous red polygons is 326 and 327. 
In qfield the two red polygons don't exist, but there is another polygon in qfield with FID 326. 
The polygon with FID 326 in qfield has FID 328 in QGIS 

The FID field has the standard setting (text widget, no NULL, no duplicates). It is editable and not hidden. I did not change any of these values.

As a test I will see if changing the FID field setting to hidden and not editable might have any efect.

Please let me know any thoughts on this issue as to why these FID values change and more importantly why the merged polyons errounnously return with overlap as result. 
 
### Reproduction steps
This problem does not occur to every split and merge action. at least two similar cases were applied correctly.

### Expected behavior
Expeted behaviour is the downloaded file should look exactly as it is on qfield app.

### Observed behavior
The downloaded files in QGIS look different. the split and merge actions were not applied correctly

### Screenshots and GIFs
[If applicable, add screenshots and GIFs to help explain your problem.]

[Please also attach additional files if a specific project/dataset is useful to investigate the problem.]

### Desktop (please complete the following information)
 - OS: Windows
 - QGIS Version  3.22.4

### Mobile (please complete the following information)
 - QField version: 3.1.9 borneo f2829b

**Comment by elmonchas** [comment]
so yesterday was another data collection day in the field. I did a few similar split and merge actions. I checked the results in qgis today and it seemed all ok this time.

There is one odd thing though. I noticed the date-time stamp ($now function and update upon modification) of the split and merge features was set on the time of pushing to qfield cloud. My guess is that with the upload to cloud, these features must have changed somehow again. Maybe another FID or FID_1 value? Other features with just split, of just changed attributes showed correct dat time stamp.

**Comment by nirvn** [comment]
@elmonchas , can we close this issue? It seems like your last comment indicates the issue described above is fixed.

**Comment by elmonchas** [comment]
Hi Mathieu,

This issue with merge function still exists in these project setups.  Because this is very problematic for our data collection, I implemented a work around solution. I am using polygon layer and split function only and flag polygons to be merged at the end of the data collection in QGIS.

I made extensive additional documentation of the issue (19-5-2024, sent through general ticket system). Probably I uploaded a new GPKG with the errors fixed, so not reproducable anymore

Here are some reproducable examples of the corruption leading to data loss.

Project on qfield cloud: Overview | LNH/vegkart_ilperveld_2425 | QFieldCloud<https://app.qfield.cloud/a/LNH/vegkart_ilperveld_2425/>

Examples of wrong changes that were caused by splitting up (more then once)  polygons, and after that deleting parts and drawing new polygons:

ID: B30a3ffc
Cloud tried to save this polygon to the wrong layer (point layer sort_invoer_v2), hence it cannot be applied

ID a20a4acf and ID 67caff40
These two ID’s ended up with the same PK 1715936536665. It results in the loss of this polygon.

I cannot rule out the possibility that this mix up of primary keys in the cloud delta file is actually also the cause of the crash when applied merge action after multiple split actions.

To summarize the problem:

  *   multiple split actions on one polygon
  *   delete some parts
  *   redraw,

  *   results in data corruption on the cloud. PK’s get mixed up.


  *   Multiple split actions on one polygon
  *   Merge parts with other existing polygons.

  *   Leads to crash of “illegal value in deltafile” possibly the same issue here


Van: Mathieu Pellerin ***@***.***>
Verzonden: zondag 11 augustus 2024 10:27
Aan: opengisch/QField ***@***.***>
CC: Wilco Non ***@***.***>; Mention ***@***.***>
Onderwerp: Re: [opengisch/QField] data corruption after synchronization to qgis (Issue #5081)


@elmonchas<https://github.com/elmonchas> , can we close this issue? It seems like your last comment indicates the issue described above is fixed.
